### PR TITLE
refactor: Optimize signal-slot connections logic

### PR DESCRIPTION
**Pull request by hebasto**
This PR:
- implements an idea from https://github.com/bitcoin/bitcoin/pull/17937#issuecomment-575991765
- simplifies `WalletFrame` class interface
- as a side effect, removes `bitcoingui` -> `walletframe` -> `bitcoingui` circular dependency
- is an alternative to https://github.com/bitcoin/bitcoin/pull/17500

### CODE REVIEW

**Comment by hebasto** [comment]
This PR was originally opened in the https://github.com/bitcoin/bitcoin repo: https://github.com/bitcoin/bitcoin/pull/17966 has a [concept ACK](https://github.com/bitcoin/bitcoin/pull/17966#issuecomment-576098433) from @jonasschnelli.

**Comment by jonasschnelli** [comment]
Concept ACK

**Comment by hebasto** [comment]
Updated 3eb6d5b4068194f3420a9f8b30cffd210ca7265b -> e7c7fbe9259c19ce55faa7397d5998b27d851224 ([pr29.01](https://github.com/hebasto/gui/commits/pr29.01) -> [pr29.02](https://github.com/hebasto/gui/commits/pr29.02)):

- rebased
- some commits are split into #333

**Comment by hebasto** [comment]
Rebased e7c7fbe9259c19ce55faa7397d5998b27d851224 -> 62cb8d98d27e7f316f01f177f35ad0ed6f8cd9ce ([pr29.02](https://github.com/hebasto/gui/commits/pr29.02) -> [pr29.03](https://github.com/hebasto/gui/commits/pr29.03)) on top of the recent CI changes.

**Review by promag — COMMENTED** [review]
Code review ACK 62cb8d98d27e7f316f01f177f35ad0ed6f8cd9ce. Will build and test.

Nice to see you picked https://github.com/bitcoin/bitcoin/pull/17500#issuecomment-575088028 suggestion.

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -680,6 +688,18 @@ void BitcoinGUI::addWallet(WalletModel* walletModel)
     }
     const QString display_name = walletModel->getDisplayName();
     m_wallet_selector->addItem(display_name, QVariant::fromValue(walletModel));
+
+    connect(wallet_view, &WalletView::outOfSyncWarningClicked, walletFrame, &WalletFrame::outOfSyncWarningClicked);
```
Comment (promag) [inline src/qt/bitcoingui.cpp:692]: nit, maybe this should stay in `WalletFrame`?

**Review by promag — COMMENTED** [review]
Tested ACK 62cb8d98d27e7f316f01f177f35ad0ed6f8cd9ce on macos 11.2.3 with depends build.

Ran with QT_FATAL_WARNINGS=0, didn't observe behavior changes.

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (hebasto) [reply src/qt/bitcoingui.cpp:692]: Doing in this way keeps compatibility with #333.

**Review by jarolrod — COMMENTED** [review]
ACK 62cb8d98d27e7f316f01f177f35ad0ed6f8cd9ce

Tested that there are no weird side-effects. Also used @promag [instructions](https://github.com/bitcoin-core/gui/pull/29#pullrequestreview-675371355) to look for behavior changes aside from manual testing. Ran on `Arch Linux Qt 5.15.2`, `macOS 11.3 Qt 5.15.2` and cross-compile to `Windows 10`.

Overall, I think the changes in this PR provide a clearer definition of the functionality the `walletframe` should be responsible for performing vs being aware of. Thanks for working on this.

**Pull request merged**